host/monitor.go

```diff
@@ -0,0 +1,213 @@
+	Enabled bool `json:"enabled,omitempty"`
+	Hosts   int  `json:"hosts,omitempty"`
+}
+
+type Monitor struct {
+	addr       string
+	dm         *DiscoverdManager
+	discoverd  *discoverdWrapper
+	discClient *discoverd.Client
+	monitorSvc discoverd.Service
+	isLeader   bool
+	c          *cluster.Client
+	hostCount  int
+	deadline   time.Time
+	shutdownCh chan struct{}
+}
+
+func NewMonitor(dm *DiscoverdManager, addr string) *Monitor {
+	return &Monitor{
+		dm:         dm,
+		discoverd:  nil,
+		addr:       addr,
+		shutdownCh: make(chan struct{}),
+	}
+}
+
+func (m *Monitor) waitDiscoverd() {
+	for {
+		if m.dm.localConnected() {
+			m.discClient = discoverd.NewClient()
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (m *Monitor) waitRaftLeader() {
+	for {
+		_, err := m.discClient.RaftLeader()
+		if err == nil {
+			break
+		}
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func (m *Monitor) waitEnabled() {
+	for {
+		monitorMeta, err := m.monitorSvc.GetMeta()
+		if err != nil {
+			time.Sleep(5 * time.Second)
+			continue
+		}
+		var decodedMeta MonitorMetadata
+		if err := json.Unmarshal(monitorMeta.Data, &decodedMeta); err != nil {
+			continue
+		}
+		if decodedMeta.Enabled {
+			m.hostCount = decodedMeta.Hosts
+			break
+		}
+		time.Sleep(5 * time.Second)
+	}
+
+}
+
+func (m *Monitor) waitRegister() {
+	for {
+		isLeader, err := m.discoverd.Register()
+		if err == nil {
+			m.isLeader = isLeader
+			break
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func (m *Monitor) Run() {
+	log := monitorLogger.New("fn", "Run")
+	log.Info("waiting for discoverd")
+	m.waitDiscoverd()
+
+	log.Info("waiting for raft leader")
+	m.waitRaftLeader()
+
+	// we can connect the leader election wrapper now
+	m.discoverd = newDiscoverdWrapper(m.addr + ":1113")
+	// connect cluster client now that discoverd is up.
+	m.c = cluster.NewClient()
+
+	m.monitorSvc = discoverd.NewService("cluster-monitor")
+
+	log.Info("waiting for monitor service to be enabled for this cluster")
+	m.waitEnabled()
+
+	log.Info("registering cluster-monitor")
+	m.waitRegister()
+
+	leaderCh := m.discoverd.LeaderCh()
+	ticker := time.NewTicker(checkInterval)
+
+	log.Info("starting monitor loop")
+	for {
+		var isLeader bool
+		select {
+		case <-m.shutdownCh:
+			log.Info("shutting down monitor")
+			return
+		case isLeader = <-leaderCh:
+			m.isLeader = isLeader
+			continue
+		default:
+		}
+
+		select {
+		case <-ticker.C:
+			if m.isLeader {
+				m.checkCluster()
+			}
+		}
+	}
+}
+
+func (m *Monitor) checkCluster() {
+	log := monitorLogger.New("fn", "checkCluster")
+	var faulted bool
+	hosts, err := m.c.Hosts()
+	if err != nil || len(hosts) < m.hostCount {
+		log.Info("waiting for hosts", "current", len(hosts), "want", m.hostCount)
+		return
+	}
+
+	controllerService := discoverd.NewService("controller")
+	controllerInstances, _ := controllerService.Instances()
+	if len(controllerInstances) == 0 {
+		log.Error("did not find any controller api instances")
+		faulted = true
```

Comment:
If discoverd is down/otherwise incapacitated won't the query to `Hosts()` result in an err and thus short circuiting the rest of the checkCluster function anyway?